enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,332 @@
+replaces:
+superseded-by:
+---
+
+# AWS EFS CSI driver operator via OLM
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Operational readiness criteria is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+AWS EFS CSI driver (+ corresponding operator) is an optional OCP component. It should be installed through OLM, when
+users opts-in.
+
+This document describes existing (unsupported) solution and how to turn it into a supported one, both from code (and
+build and shipment), and from user perspective.
+
+## Motivation
+
+Some of our customers want to use Amazon Elastic FileSystem in their clusters. At the same time, not all customers want
+to use it, so it should not be installed by default as our AWS EBS CSI driver.
+
+Despite written by Red Hat employees, the existing CSI driver operator
+[aws-efs-operator](https://github.com/openshift/aws-efs-operator) uses upstream CSI driver, upstream CSI sidecars and
+[the operator itself](https://quay.io/repository/app-sre/aws-efs-operator?tab=tags) is not released through
+registry.redhat.com and therefore cannot be supported by us.
+
+### Goals
+
+* Allow users to use AWS EFS volumes in a supported way.
+* Allow existing users of community-supported aws-efs-operator to update to Red Hat supported solution (mostly manually).
+* Keep existing community-supported aws-efs-operator features (namely, `SharedVolume` in
+  `aws-efs.managed.openshift.io` API group). See below.
+
+### Non-Goals
+
+* Change the EFS CSI driver in any way. All missing functionality there must go through RFE process.
+
+## Existing situation
+
+### [aws-efs-operator](https://github.com/openshift/aws-efs-operator)
+
+Summary of current [aws-efs-operator](https://github.com/openshift/aws-efs-operator) (no judging, just stating the facts):
+
+* It's written using `controller-runtime`.
+* Once the operator is installed by OLM, the operator automatically installs the CSI driver, without any CRD/CR.
+  * As consequence, user cannot un-install the CSI driver easily.
+* By default, it is installed in `openshift-operators` namespace and the CSI driver runs there too.
+* It does not create EFS volumes in AWS! It's up to the cluster admin to create EFS volume in AWS, figure in which VPC
+  it should be available and set correct security group. See [this KB article](https://access.redhat.com/articles/5025181)
+  for suggested procedure.
+* It offers `SharedVolume` CRD in `aws-efs.managed.openshift.io/v1alpha1` API group to create PV + PVC for an EFS
+  volume. The EFS volume still needs to be created manually, see the previous bullet!
+  * All that `SharedVolume` really does is that it creates PV with proper settings + a PVC. This allows unprivileged
```

Comment:
I found this very unsual and different to other storage providers I've used before. For example I'm installing a helm chart which creates a RWX PVC - I'd expect this to be dynamically provisioned and work. SharedVolume breaks this workflow.

Reply:
This has been discussed several times and each time weighs the scales more heavily in favor of not carrying the `SharedVolume` CR/concept forward into this solution.

(However, I'll point out that, even if `SharedVolume` _is_ implemented, it wouldn't be required as part of the workflow. You could ignore it and create your PVs and PVCs directly in the same way you're used to.)